components/scream/src/physics/zm/atmosphere_deepconvection.cpp

```diff
@@ -0,0 +1,165 @@
+#include "physics/zm/scream_zm_interface.hpp"
+#include "physics/zm/atmosphere_deepconvection.hpp"
+#include "physics/zm/zm_inputs_initializer.cpp"
+
+namespace scream
+
+{
+
+ZMDeepConvection::ZMDeepConvection (const Comm& comm,const ParameterList& /* params */)
+  : m_zm_comm (comm)
+{
+  m_initializer = create_field_initializer<ZMInputsInitializer>();
+  
+}
+
+
+void ZMDeepConvection::set_grids(const std::shared_ptr<const GridsManager> grids_manager)
+{
+  using namespace std;
+
+  using namespace units;
+  auto Q = kg/kg;
+  auto nondim = m/m;
+  Q.set_string("kg/kg");
+
+  constexpr int NVL = 72;  /* TODO THIS NEEDS TO BE CHANGED TO A CONFIGURABLE */
+  constexpr int QSZ =  35;  /* TODO THIS NEEDS TO BE CHANGED TO A CONFIGURABLE */
+  auto grid = grids_manager->get_grid("Physics");
+  const int num_dofs = grid->get_num_local_dofs();
+  const int nc = num_dofs;
+
+  using namespace ShortFieldTagsNames;
+
+  FieldLayout scalar3d_layout_mid { {COL,VL}, {nc,NVL} }; // Note that C++ and Fortran read array dimensions in reverse
+  FieldLayout scalar3d_layout_int { {COL,VL}, {nc,NVL+1} }; // Note that C++ and Fortran read array dimensions in reverse
+  FieldLayout vector3d_layout_mid{ {COL,CMP,VL}, {nc,QSZ,NVL} };
+  FieldLayout tracers_layout { {COL,VAR,VL}, {nc,QSZ,NVL} };
+  FieldLayout scalar2d_layout{ {COL}, {nc} };
+ 
+  std::vector<FieldLayout> layout_opts = {scalar3d_layout_mid, scalar3d_layout_int,
+					vector3d_layout_mid, tracers_layout, scalar2d_layout};
+ 
+  set_grid_opts();
+  
+  for ( auto i = opt_map.begin(); i != opt_map.end(); ++i) {
+    m_required_fields.emplace((i->second).name, layout_opts[((i->second).field_idx)], Q, grid->name());
+    if ( (i->second).isOut == true ) {
+      m_computed_fields.emplace((i->second).name, layout_opts[((i->second).field_idx)], Q, grid->name());
+    }
+  }
+  
+}
+
+// =========================================================================================
+void ZMDeepConvection::initialize (const util::TimeStamp& t0)
+{
+  m_current_ts = t0;
+  
+  zm_init_f90 (*m_raw_ptrs_in["limcnv_in"], m_raw_ptrs_in["no_deep_pbl_in"]);
+  
+  using strvec = std::vector<std::string>;
+  const strvec& initable = zm_inputs;
+  
+  for (const auto& name : initable) {
```

Comment:
You need to think about the initialization, based on what outcome you want. As you implemented it, if any field that ZM needs is not inited, ZM will init it. One thing you may want to check is whether you're ok with ZM initing a few of the inputs, but not all. Like, if all but one fields already have an initialization, is that ok or is it a symptom of an error (like, that field was supposed to be inited alongside the others)?

You can look at P3, for an example on how to guard from that scenario.

Reply:
Good call. Working on implementing similar checks as P3 has.